### PR TITLE
Incremental EDS - merge of 8722 from 1.0

### PR DESCRIPTION
**Pull request by costinm**
This is a manual merge, the trees have diverged too much. 

I did not merge the 'legacy' code or the variables to control fallback, not needed in master.

From original PR:
- added a mechanism for service discovery to update 'shards' of endpoints
- new data structures optimized for endpoint update ( IstioEndpoint has minimum info needed and allows caching the proto )
- optimized the update path - only affected shard will be changed without triggering a full cache invalidation
- rewrote the EDS tests - which were testing the 0.7 method, no longer used
- added a ADS client mirroring envoy behavior, to simplify testing (the load tester using this will be in separate PR)
- optimized the incremental push update path.
- fixed label update to cause a push.

### CODE REVIEW

File: pilot/pkg/proxy/envoy/v2/eds.go [hunk 1]
```diff
@@ -142,6 +185,171 @@ func newEndpoint(e *model.NetworkEndpoint) (*endpoint.LbEndpoint, error) {
...
+	}
+	return nil
+}
+
+// updateServiceShards will list the endpoints and create the shards.
+// This is used to reconcile and to support non-k8s registries (until they migrate).
+// Note that aggreaged list is expensive (for large numbers) - we want to replace
+// it with a model where DiscoveryServer keeps track of all endpoint registries
+// directly, and calls them one by one.
+func (s *DiscoveryServer) updateServiceShards(push *model.PushContext) error {
```
Comment (rshriram) [inline pilot/pkg/proxy/envoy/v2/eds.go]: This is endpointshards isn’t it?

Comment (costinm) [reply pilot/pkg/proxy/envoy/v2/eds.go]: Right, will update

**Comment by costinm** [comment]
pkg/adsc is covered by the tests in pilot/proxy/envoy/v2 - yes, we'll probably move some of the tests to the adsc package. It used to be part of the tests - but now we use it in the load tester.

**Pull request merged**